apps/src/aichat/redux/aichatRedux.ts

```diff
@@ -1,35 +1,93 @@
-import {createSlice, PayloadAction} from '@reduxjs/toolkit';
-import {ChatCompletionMessage} from '../types';
+import {createSlice, PayloadAction, createAsyncThunk} from '@reduxjs/toolkit';
+import {
+  ChatCompletionMessage,
+  AichatLevelProperties,
+  Status,
+  Role,
+} from '../types';
 import {initialChatMessages} from '../constants';
 
 const registerReducers = require('@cdo/apps/redux').registerReducers;
+import {LabState} from '@cdo/apps/lab2/lab2Redux';
+import {getChatCompletionMessage} from '../chatApi';
 
 export interface AichatState {
   // All user and assistant chat messages - includes too personal and inappropriate user messages.
   // Messages will be logged and stored.
   chatMessages: ChatCompletionMessage[];
-  // New user message that is submitted via the user chat message editor.
-  newUserMessage: string;
   // Denotes whether we are waiting for a chat completion response from the backend
   isWaitingForChatResponse: boolean;
   // Denotes whether we should show the warning modal
   showWarningModal: boolean;
 }
 
 const initialState: AichatState = {
-  newUserMessage: '',
   chatMessages: initialChatMessages,
   isWaitingForChatResponse: false,
   showWarningModal: true,
 };
 
+// THUNKS
+export const submitChatMessage = createAsyncThunk(
```

Comment:
Great comments throughout this function! Can we add a top-level one here as well just as an overview? Probably something like `// submits a chat message to the chat completion endpoint, waits for a response, and updates message state`

Reply:
Yes - definitely can add: Thanks!
```
// This thunk's callback function submits a user chat message to the chat completion endpoint,
// waits for a chat completion response, and updates the user message state.
```